### ISSUE TITLE
Building Tensorflow Examples via Bazel on macOS doesn't work. Building for Object Tracking Support.

### ISSUE THREAD

**Issue by Zeit42** [tensorflow/tensorflow#27199]
<em>Please make sure that this is a build/installation issue. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:build_template</em>

**System information**
- OS Platform and Distribution: macOS 10.13.6
- Mobile device: Not applicable
- TensorFlow installed from (source or binary): Source
- TensorFlow version: The latest one as of writing
- Python version: 2.7
- Installed using virtualenv? pip? conda?: no
- Bazel version (if compiling from source): 0.23.2
- GCC/Compiler version (if compiling from source): 10.0.0
- CUDA/cuDNN version: none
- GPU model and memory: none

**Describe the problem**

I'm trying to build the Tensorflow Examples via Bazel so that I can have the Object Tracking Support for Object Detection in TF-Lite.

**Provide the exact sequence of commands / steps that you executed before running into the problem**

[I used this guide](https://github.com/tensorflow/tensorflow/tree/master/tensorflow/lite/examples/android/app#building-from-source-with-bazel), supplemented with [this guide](https://github.com/tensorflow/tensorflow/tree/master/tensorflow/examples/android#bazel)

And here are the steps I took.

> cd ~/my/android/directory/
>git clone --recurse-submodules https://github.com/tensorflow/tensorflow.git

I checked for my bazel version:
> bazel version
> WARNING: --batch mode is deprecated. Please instead explicitly shut down your Bazel server using the command "bazel shutdown".
> Build label: 0.23.2
> Build target: bazel-out/darwin-opt/bin/src/main/java/com/google/devtools/build/lib/bazel/BazelServer_deploy.jar
> Build time: Mon Mar 11 16:50:03 2019 (1552323003)
> Build timestamp: 1552323003
> Build timestamp as int: 1552323003

> cd tensorflow
> ./configure

Here were the output, I just hit the defaults for all of these:

> WARNING: --batch mode is deprecated. Please instead explicitly shut down your Bazel server using the command "bazel shutdown".
> You have bazel 0.23.2 installed.
> Please specify the location of python. [Default is /usr/local/opt/python@2/bin/python2.7]: 

> Found possible Python library paths:
>   /usr/local/Cellar/python@2/2.7.15_1/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages
> Please input the desired Python library path to use.  Default is [/usr/local/Cellar/python@2/2.7.15_1/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages]

> Do you wish to build TensorFlow with XLA JIT support? [y/N]: 
> No XLA JIT support will be enabled for TensorFlow.

> Do you wish to build TensorFlow with OpenCL SYCL support? [y/N]: 
> No OpenCL SYCL support will be enabled for TensorFlow.

> Do you wish to build TensorFlow with ROCm support? [y/N]: 
> No ROCm support will be enabled for TensorFlow.

> Do you wish to build TensorFlow with CUDA support? [y/N]: 
> No CUDA support will be enabled for TensorFlow.

> Do you wish to download a fresh release of clang? (Experimental) [y/N]: 
> Clang will not be downloaded.

> Do you wish to build TensorFlow with MPI support? [y/N]: 
> No MPI support will be enabled for TensorFlow.

> Please specify optimization flags to use during compilation when bazel option "--config=opt" is specified [Default is -march=native -Wno-sign-compare]: 

> Would you like to interactively configure ./WORKSPACE for Android builds? [y/N]: 
> Not configuring the WORKSPACE for Android builds.

> Would you like to configure TensorFlow for iOS builds? [y/N]: 
> Not configuring TensorFlow for iOS builds.

> Preconfigured Bazel build configs. You can use any of the below by adding "--config=<>" to your build command. See .bazelrc for more details.
> 	--config=mkl         	# Build with MKL support.
> 	--config=monolithic  	# Config for mostly static monolithic build.
> 	--config=gdr         	# Build with GDR support.
> 	--config=verbs       	# Build with libverbs support.
> 	--config=ngraph      	# Build with Intel nGraph support.
> 	--config=numa        	# Build with NUMA support.
> 	--config=dynamic_kernels	# (Experimental) Build kernels into separate shared objects.
> Preconfigured Bazel build configs to DISABLE default on features:
> 	--config=noaws       	# Disable AWS S3 filesystem support.
> 	--config=nogcp       	# Disable GCP support.
> 	--config=nohdfs      	# Disable HDFS support.
> 	--config=noignite    	# Disable Apache Ignite support.
> 	--config=nokafka     	# Disable Apache Kafka support.
> 	--config=nonccl      	# Disable NVIDIA NCCL support.
> Configuration finished

The guide in the 2nd link says:

> "The Android entries in <workspace_root>/WORKSPACE must be uncommented with the paths filled in appropriately depending on where you installed the NDK and SDK. Otherwise an error such as: "The external label '//external:android/sdk' is not bound to anything" will be reported."

But I don't see anything on that directory aside from the tensorflow folder that was cloned?

And in the next paragraph, it says:
> "Also edit the API levels for the SDK in WORKSPACE to the highest level you have installed in your SDK. This must be >= 23 (this is completely independent of the API level of the demo, which is defined in AndroidManifest.xml). The NDK API level may remain at 14."

But when I checked my WORKSPACE file through nano and cat, I only see this:

> workspace(name = "org_tensorflow")
> 
> load("@bazel_tools//tools/build_defs/repo:http.bzl", "http_archive", "http_file")
> 
> http_archive(
>     name = "io_bazel_rules_closure",
>     sha256 = "ddce3b3a3909f99b28b25071c40b7fec7e2e1d1d1a4b2e933f3082aa99517105",
>     strip_prefix = "rules_closure-316e6133888bfc39fb860a4f1a31cfcbae485aef",
>     urls = [
>         "https://mirror.bazel.build/github.com/bazelbuild/rules_closure/archive/316e6133888bfc39fb860a4f1a31cfcbae485aef.tar.gz",
>         "https://github.com/bazelbuild/rules_closure/archive/316e6133888bfc39fb860a4f1a31cfcbae485aef.tar.gz",  # 2019-03-21
>     ],
> )
> 
> load("@io_bazel_rules_closure//closure:defs.bzl", "closure_repositories")
> 
> closure_repositories()
> 
> load("//third_party/toolchains/preconfig/generate:archives.bzl",
>      "bazel_toolchains_archive")
> 
> bazel_toolchains_archive()
> 
> load(
>     "@bazel_toolchains//repositories:repositories.bzl",
>     bazel_toolchains_repositories = "repositories",
> )
> 
> bazel_toolchains_repositories()
> 
> load(
>     "@io_bazel_rules_docker//repositories:repositories.bzl",
>     container_repositories = "repositories",
> )
> 
> container_repositories()
> 
> load("//third_party/toolchains/preconfig/generate:workspace.bzl",
>      "remote_config_workspace")
> 
> remote_config_workspace()
> 
> # Apple and Swift rules.
> http_archive(
>     name = "build_bazel_rules_apple",
>     sha256 = "4b90786009fa8df25230442244bad2832ba8d6bc4987f68150a7de59c8827e90",
>     strip_prefix = "rules_apple-0.14.0",
>     urls = ["https://github.com/bazelbuild/rules_apple/archive/0.14.0.tar.gz"],
> )
> http_file(
>    name = "xctestrunner",
>     executable = 1,
>     urls = ["https://github.com/google/xctestrunner/releases/download/0.2.6/ios_test_runner.par"],
> )
> 
> http_archive(
>     name = "bazel_skylib",
>     sha256 = "2c62d8cd4ab1e65c08647eb4afe38f51591f43f7f0885e7769832fa137633dcb",
>     strip_prefix = "bazel-skylib-0.7.0",
>     urls = ["https://github.com/bazelbuild/bazel-skylib/archive/0.7.0.tar.gz"],
> )
> 
> http_archive(
>     name = "build_bazel_apple_support",
>     sha256 = "835663c4bb02f4bf01dce8a2a176df7fa682dbb867d3698ae12258c1628bb8f0",
>     strip_prefix = "apple_support-0.5.0",
>     urls = ["https://github.com/bazelbuild/apple_support/archive/0.5.0.tar.gz"],
> )
> 
> http_archive(
>     name = "build_bazel_rules_swift",
>     sha256 = "32d124878cd49775d84f59ba90440c8b23b7c775aec8fec1978f751c76ddee8a",
>     strip_prefix = "rules_swift-0.7.0",
>     urls = ["https://github.com/bazelbuild/rules_swift/archive/0.7.0.tar.gz"],
> )
> 
> http_archive(
>     name = "com_github_apple_swift_swift_protobuf",
>     type = "zip",
>     strip_prefix = "swift-protobuf-1.2.0/",
>     urls = ["https://github.com/apple/swift-protobuf/archive/1.2.0.zip"],
> )
> 
> # Use swift_rules_dependencies to fetch the tolchains.
> # Since we defined all the "git_repository" rules above, the following call will
> # skip redefining them.
> load("@build_bazel_rules_swift//swift:repositories.bzl", "swift_rules_dependencies")
> swift_rules_dependencies()
> 
> # We must check the bazel version before trying to parse any other BUILD
> # files, in case the parsing of those build files depends on the bazel
> # version we require here.
> load("//tensorflow:version_check.bzl", "check_bazel_version_at_least")
> check_bazel_version_at_least("0.19.0")
> 
> load("//tensorflow:workspace.bzl", "tf_workspace")
> 
> load("//third_party/android:android_configure.bzl", "android_configure")
> android_configure(name="local_config_android")
> load("@local_config_android//:android.bzl", "android_workspace")
> android_workspace()
> 
> # Please add all new TensorFlow dependencies in workspace.bzl.
> tf_workspace()
> 
> http_archive(
>     name = "inception_v1",
>     build_file = "//:models.BUILD",
>     sha256 = "7efe12a8363f09bc24d7b7a450304a15655a57a7751929b2c1593a71183bb105",
>     urls = [
>         "http://storage.googleapis.com/download.tensorflow.org/models/inception_v1.zip",
>         "http://download.tensorflow.org/models/inception_v1.zip",
>     ],
> )
> 
> http_archive(
>     name = "mobile_ssd",
>     build_file = "//:models.BUILD",
>     sha256 = "bddd81ea5c80a97adfac1c9f770e6f55cbafd7cce4d3bbe15fbeb041e6b8f3e8",
>     urls = [
>         "http://storage.googleapis.com/download.tensorflow.org/models/object_detection/ssd_mobilenet_v1_android_export.zip",
>         "http://download.tensorflow.org/models/object_detection/ssd_mobilenet_v1_android_export.zip",
>     ],
> )
> 
> http_archive(
>    name = "mobile_multibox",
>     build_file = "//:models.BUILD",
>     sha256 = "859edcddf84dddb974c36c36cfc1f74555148e9c9213dedacf1d6b613ad52b96",
>     urls = [
>         "http://storage.googleapis.com/download.tensorflow.org/models/mobile_multibox_v1a.zip",
>         "http://download.tensorflow.org/models/mobile_multibox_v1a.zip",
>     ],
> )
> 
> http_archive(
>     name = "stylize",
>     build_file = "//:models.BUILD",
>     sha256 = "3d374a730aef330424a356a8d4f04d8a54277c425e274ecb7d9c83aa912c6bfa",
>     urls = [
>         "http://storage.googleapis.com/download.tensorflow.org/models/stylize_v1.zip",
>         "http://download.tensorflow.org/models/stylize_v1.zip",
>     ],
> )
> 
> http_archive(
>     name = "speech_commands",
>     build_file = "//:models.BUILD",
>     sha256 = "c3ec4fea3158eb111f1d932336351edfe8bd515bb6e87aad4f25dbad0a600d0c",
>     urls = [
>         "http://storage.googleapis.com/download.tensorflow.org/models/speech_commands_v0.01.zip",
>         "http://download.tensorflow.org/models/speech_commands_v0.01.zip",
>     ],
> )

There wasn't anything that resembled `api_levels` or anything similar. 

And when I finally hit:
> bazel build -c opt --cxxopt='--std=c++11' --fat_apk_cpu=x86,x86_64,arm64-v8a,armeabi-v7a \
>  //tensorflow/lite/examples/android:tflite_demo

I got the following errors:

> ERROR: /private/var/tmp/_bazel_angelopvillasanta/70db096b7d05a90f0f37f9f041e057b0/external/local_config_cc/BUILD:45:1: in cc_toolchain_suite rule @local_config_cc//:toolchain: cc_toolchain_suite '@local_config_cc//:toolchain' does not contain a toolchain for cpu 'arm64-v8a'
> ERROR: Analysis of target '//tensorflow/lite/examples/android:tflite_demo' failed; build aborted: Analysis of target '@local_config_cc//:toolchain' failed; build aborted
> INFO: Elapsed time: 14.026s
> INFO: 0 processes.
> FAILED: Build did NOT complete successfully (24 packages loaded, 215 targets configured)
>     Fetching @local_jdk; fetching

Now I'm not sure what I should do next. I'm not sure how to install C++11 on the macOS, as that seems to be the only dependency I'm not sure I have covered. 

Basically my objective is to build the Tensorflow example in a macOS, because I need to be able to have support of the Object Tracking feature in Object Detection for Tensorflow-Lite.

Is this possible? Did I miss any steps?

**Any other info / logs**
The logs are all above.

**Comment by patrick-ucr** [comment]
I made this with cmake.

1. Add this to build gradle
`defaultConfig {
        ndk {
            abiFilters "${cpuType}"
        }
        externalNativeBuild {
            cmake {
                cppFlags "-std=c++11", "-Wall"
                arguments '-DANDROID_TOOLCHAIN=clang', '-DANDROID_STL=c++_static'
            }
        }
    }
    externalNativeBuild {
        cmake {
            path './jni/CMakeLists.txt'
        }
    }`

2. Copy the jni folder from tensorflow/tensorflow/example/android that contains code for object tracking

3. Make sure that the paths inside CMakeLists are consistent with your environment.

Now you can have object tracking right out of Android Studio without having to deal with complicated Bazel build.

**Comment by Zeit42** [comment]
> I made this with cmake.
> 
> 1. Add this to build gradle
>    `defaultConfig { ndk { abiFilters "${cpuType}" } externalNativeBuild { cmake { cppFlags "-std=c++11", "-Wall" arguments '-DANDROID_TOOLCHAIN=clang', '-DANDROID_STL=c++_static' } } } externalNativeBuild { cmake { path './jni/CMakeLists.txt' } }`
> 2. Copy the jni folder from tensorflow/tensorflow/example/android that contains code for object tracking
> 3. Make sure that the paths inside CMakeLists are consistent with your environment.
> 
> Now you can have object tracking right out of Android Studio without having to deal with complicated Bazel build.

Did you have to download the CMake module via Android Studio's SDK Manager Tool?

![Screen Shot 2019-03-30 at 6 17 03 PM](https://user-images.githubusercontent.com/1715673/55274839-0b92c780-5318-11e9-833e-2c2544690245.png)

**Comment by patrick-ucr** [comment]
> > I made this with cmake.
> > 
> > 1. Add this to build gradle
> >    `defaultConfig { ndk { abiFilters "${cpuType}" } externalNativeBuild { cmake { cppFlags "-std=c++11", "-Wall" arguments '-DANDROID_TOOLCHAIN=clang', '-DANDROID_STL=c++_static' } } } externalNativeBuild { cmake { path './jni/CMakeLists.txt' } }`
> > 2. Copy the jni folder from tensorflow/tensorflow/example/android that contains code for object tracking
> > 3. Make sure that the paths inside CMakeLists are consistent with your environment.
> > 
> > Now you can have object tracking right out of Android Studio without having to deal with complicated Bazel build.
> 
> Did you have to download the CMake module via Android Studio's SDK Manager Tool?
> 
> ![Screen Shot 2019-03-30 at 6 17 03 PM](https://user-images.githubusercontent.com/1715673/55274839-0b92c780-5318-11e9-833e-2c2544690245.png)

Yes I did.

**Comment by jdduke** [comment]
> Would you like to interactively configure ./WORKSPACE for Android builds? [y/N]:
> Not configuring the WORKSPACE for Android builds.

When you run ./configure you need to enable it for Android builds.

**Comment by jvishnuvardhan** [comment]
@Zeit42 Please close the issue if the issue was resolved. Thanks!

**Comment by jvishnuvardhan** [comment]
Closing this out since I understand it to be resolved, but please let me know if I'm mistaken.
